feat: psp22 trait contract example with drink tests (#300)
r0gue-io/pop-node: pop-api/examples/fungibles/README.md

As the token owner, the contract has permissions to mint and burn tokens, but it can only transfer and approve tokens on its own behalf and requires explicit approval to transfer tokens for other accounts. This structure enables seamless integration and interoperability across the Polkadot ecosystem and its applications.

## Use Cases

This contract can serve a variety of purposes where owner-controlled token management is essential. Example use cases include:
- **DAO Token**: A DAO can use this contract to manage a governance token, with the DAO overseeing token issuance and removal based on governance decisions.
- **Staking and Rewards**: This contract supports minting tokens specifically for reward distribution.
- **Loyalty Programs**: Businesses or platforms can use this contract to issue loyalty points, with the owner managing token balances for users based on participation or purchases.ints and burns tokens for a staking rewards program, allowing the staking authority to distribute rewards only to eligible participants.

## Test with Pop Drink

> Since this contract interacts directly with Pop’s runtime through the Pop API, it requires [Pop Drink](https://github.com/r0gue-io/pop-drink) for testing. See how the contract is tested in [tests](./tests.rs).

Reviewer: ```suggestion
Since this contract interacts directly with Pop’s runtime through the Pop API, it requires [Pop Drink](https://github.com/r0gue-io/pop-drink) for testing. See how the contract is tested in [tests](./tests.rs).

```

# PSP22 Fungible Token with Pop API

> This [ink!][ink] contract implements a [PSP22-compliant][psp22] fungible token by leveraging the [Pop API Fungibles][pop-api-fungibles]. Unlike typical token contracts, where the contract itself manages the token, tokens created by this contract are managed directly by Pop. This enables seamless integration and interoperability of the token across the Polkadot ecosystem and its applications. 

Author: Great feedback. This makes sense to me. I just added a few key bullet points. Let me know if you have any other bullet points that you are happy to add.

> As the creator of the token, the contract has permissions to mint and burn tokens, but it can only transfer and approve tokens on its own behalf and requires explicit approval to transfer tokens for other accounts. Instead of users interacting with the contract to handle their tokens, they interact primarily with Pop’s runtime. 

Reviewer 2: ```suggestion
As the creator of the token, the contract has permissions to mint and burn tokens, but it can only transfer and approve tokens on its own behalf and requires explicit approval to transfer tokens for other accounts. Instead of users interacting with the contract to handle their token approvals, they interact primarily with Pop’s runtime. 
```

## Use Cases

This contract can serve a variety of purposes where owner-controlled token management is essential. Example use cases include:
- **DAO Token**: A DAO can use this contract to manage a governance token, with the DAO overseeing token issuance and removal based on governance decisions.
- **Staking and Rewards**: This contract supports minting tokens specifically for reward distribution.

> - **Loyalty Programs**: Businesses or platforms can use this contract to issue loyalty points, with the owner managing token balances for users based on participation or purchases.ints and burns tokens for a staking rewards program, allowing the staking authority to distribute rewards only to eligible participants.

Reviewer 2: This was a guess, feel free to correct as intended.

# PSP22 Fungible Token with Pop API

This [ink!][ink] contract shows a contract that allows interaction and management of a fungible token following the [PSP22 standard][psp22], utilizing the [Pop API Fungibles][pop-api-fungibles] feature. In this specific contract only the contract owner has a permission to call specific methods.

## Design Goals

- Token exists as an asset on the Pop Network, rather than being confined solely to the contract.
- Contract is the `origin` of the calls made by `pop-api`.
- Only the contract owner has a permission to call specific methods.

> [Learn more how Pop API works.](/pop-api/README.md)

Reviewer: Remove this paragraph, design goals is not really suitable for this readme. Have we designed anything you think :) ?

> ## Test with Pop DR!nk
> 
> Because the contract uses `pop-api` which calls to the runtime, it requires a special crate called [pop-drink][pop-drink] to test the contract. See how the contract is tested in [tests](./tests.rs).

Reviewer: ```suggestion
## Test with Pop Drink

Since this contract interacts directly with Pop’s runtime through the Pop API, it requires [Pop Drink](https://github.com/r0gue-io/pop-drink) for testing. See how the contract is tested in [tests](./tests.rs).
```

> ## What can be improved?

Reviewer: ```suggestion
## Potential Improvements
```

- **Multiple owner management**: Instead of restricting ownership to a single `owner`, the contract could be designed to accommodate multiple owners.

> ## Use cases
> 
> This contract can be used in multiple different real world cases such as:
> 
> - **Governance Token in DAO**: The DAO uses this PSP22 contract to create a governance token, with DAO authority as owner, enabling actions like minting and burning tokens to manage member governance.
> - **Staking Rewards**: This contract mints and burns tokens for a staking rewards program, allowing the staking authority to distribute rewards only to eligible participants.

Reviewer: ```suggestion
## Example Use Cases

This contract can serve a variety of purposes where owner-controlled token management is essential. Example use cases include:

- **DAO Token**: A DAO can use this contract to manage a governance token, with the DAO overseeing token issuance and removal based on governance decisions.
- **Staking and Rewards**: This contract supports minting tokens specifically for reward distribution.
- **Loyalty Programs**: Businesses or platforms can use this contract to issue loyalty points, with the owner managing token balances for users based on participation or purchases.
```

Reviewer: Can we move this section above the `Tested with Pop Drink` section`.

# PSP22 Fungible Token with Pop API

PSP22 is a fungible token standard for WebAssembly smart contracts running on blockchains based on the [Substrate][substrate] framework. It is an equivalent of Ethereum's [ERC-20][erc20]. The definition of the PSP22 standard can be found [here][psp22].

> This repository contains a simple, minimal implementation of the PSP22 token in [ink!][ink] smart contract programming language (EDSL based on Rust), utilizing the [Pop API Fungibles][pop-api-fungibles] feature.

Reviewer: Way too much detail, also we don't want to repeat the docs in the pop api about the psp22 contract. Just explain the contract in a simple way :), no need to make it fancy or anything.

"This contract shows a contract that allows interaction and management of a fungible token following the psp22 standard. In this specific contract only the contract owner has a permission to call specific methods. This is because ....
This is only one way of how the fungibles api can be utilised, other ways could be ...

Tests are written with pop drink ... and can be run with `cargo test --release`

## Support

Be part of our passionate community of Web3 builders. [Join our Telegram](https://t.me/onpopio)!

Feel free to raise issues if anything is unclear, have ideas or want to contribute to Pop! We would be very happy to see other examples utilising the fungibles API!

For any questions related to ink! you can also go to [Polkadot Stack Exchange](https://polkadot.stackexchange.com/) or
ask the [ink! community](https://t.me/inkathon/1).

Author: The level of details in the current `README.md` is not that much, and I think we should give them some details and reference link for them to learn more. For example:

> Tests are written with pop drink ... and can be run with cargo test --release

Developer will have a concern what is pop-drink and why they need to use it instead of ink e2e test. So at least let them know that we need pop-drink because pop-api...then reference link

- Contract is the `origin` of the calls made by `pop-api`.
- Only the contract owner has a permission to call specific methods.

[Learn more how Pop API works.](/pop-api/README.md)

## Test with Pop DR!nk

Because the contract uses `pop-api` which calls to the runtime, it requires a special crate called [pop-drink][pop-drink] to test the contract. See how the contract is tested in [tests](./tests.rs).

## What can be improved?

- **Multiple owner management**: Instead of restricting ownership to a single `owner`, the contract could be designed to accommodate multiple owners.

## Use cases

This contract can be used in multiple different real world cases such as:

- **Governance Token in DAO**: The DAO uses this PSP22 contract to create a governance token, with DAO authority as owner, enabling actions like minting and burning tokens to manage member governance.
- **Staking Rewards**: This contract mints and burns tokens for a staking rewards program, allowing the staking authority to distribute rewards only to eligible participants.

## Support

Be part of our passionate community of Web3 builders. [Join our Telegram](https://t.me/onpopio)!

> Feel free to raise issues if anything is unclear, have ideas or want to contribute to Pop!

Reviewer: ```suggestion
Feel free to raise issues if anything is unclear, have ideas or want to contribute to Pop! Examples using the fungibles API are always welcome!
```

## Use cases

This contract can be used in multiple different real world cases such as:

- **Governance Token in DAO**: The DAO uses this PSP22 contract to create a governance token, with DAO authority as owner, enabling actions like minting and burning tokens to manage member governance.
- **Staking Rewards**: This contract mints and burns tokens for a staking rewards program, allowing the staking authority to distribute rewards only to eligible participants.

## Support

Be part of our passionate community of Web3 builders. [Join our Telegram](https://t.me/onpopio)!

Feel free to raise issues if anything is unclear, have ideas or want to contribute to Pop!

For any questions related to ink! you can also go to [Polkadot Stack Exchange](https://polkadot.stackexchange.com/) or
ask the [ink! community](https://t.me/inkathon/1).

[ink]: https://use.ink
[substrate]: https://substrate.io
[erc20]: https://ethereum.org/en/developers/docs/standards/tokens/erc-20/
[psp22]: https://github.com/inkdevhub/standards/blob/master/PSPs/psp-22.md
[pop-api-fungibles]: https://github.com/r0gue-io/pop-node/tree/main/pop-api/src/v0/fungibles

> [pallet-assets]: https://crates.io/crates/pallet-assets

Reviewer: Are we not trying to prevent the devs to have to learn about pallet assets?
